apis/src/main/java/org/apache/rocketmq/apis/consumer/PullConsumer.java

```diff
@@ -0,0 +1,172 @@
+     * @param messageQueues are the collection for current consumer.
+     * @throws ClientException when assign
+     */
+    void assign(Collection<MessageQueue> messageQueues) throws ClientException;
+
+    /**
+     * Pull consumer query and update metadata about message queues periodically, listener is triggered once metadata
+     * is updated. The listener is required only in manual assignment mode.
+     * When use the subscription mode, no need to care the messageQueue change events.
+     *
+     * @param topic    topic to query and update metadata.
+     * @param listener listener to receive changes of metadata by topic.
+     */
+    void registerMessageQueuesChangeListener(String topic, MessageQueuesChangeListener listener);
+
+    /**
+     * Add subscription expression dynamically when use subscription mode.
+     *
+     * <p>If first {@link SubscriptionExpression} that contains topicA and tag1 is exists already in consumer, then
+     * second {@link SubscriptionExpression} which contains topicA and tag2, <strong>the result is that the second one
+     * replaces the first one instead of integrating them</strong>.
+     *
+     * @param subscriptionExpression new subscription expression to add.
+     * @return pull consumer instance.
+     */
+    PullConsumer subscribe(SubscriptionExpression subscriptionExpression) throws ClientException;
+
+    /**
+     * Remove subscription expression dynamically by topic.
+     *
+     * <p>Nothing occurs if the specified topic does not exist in subscription expressions of pull consumer.
+     *
+     * @param topic the topic to remove subscription.
+     * @return pull consumer instance.
+     */
+    PullConsumer unsubscribe(String topic) throws ClientException;
+
+    /**
+     * Get the collection of messageQueues currently assigned to current consumer.
+     * @return the collection of messageQueues currently assigned to current consumer
+     */
+    Collection<MessageQueue> assignment();
+
+    /**
+     * Fetch messages from server synchronously. This method returns immediately if there are messages available.
+     * Otherwise, it will await the passed timeout. If the timeout expires, an empty map will be returned.
+     * An error occurs if you do not subscribe or assign messageQueues before polling for data.
+     * @param messageQueue the target messageQueue to pull message.
+     * @param maxMessageNum max message num when server returns.
+     * @return collection of messageViews of this queue.
+     */
+    Collection<MessageView> pull(MessageQueue messageQueue, int maxMessageNum) throws ClientException;
+
+    /**
+     * Commit the offsets for the specified messageQueue.
+     * @param messageQueue the specified messageQueue to commit offset
+     * @param committedOffset the specified offset commit to server
+     */
+    void commit(MessageQueue messageQueue, long committedOffset) throws ClientException;
```

Comment:
The commit needs an async method.

And more importantly, it needs a simple commit().

In most cases, the user only wants to control the commit time but doesn't want to manage the offset store.

Reply:
Maybe we need to rethink of the design of PullConsumer.